### DISCUSSION TITLE
Possible to define a functionality that allows for custom defaults/generation

### DISCUSSION THREAD

**Discussion by michael-jaquier** [tokio-rs/prost#1131] [Q&A]
We have a litany of Protobufs. I would like the following to function
```rust
let mut rng = rand::thread_rng();
let my_proto: MyProto = MyProto.generate(&mut rng);
let my_proto_but_different: MyProto = MyProto.generate(&mut rng);
/// ... Do work with the proto -> Send it into system for smoke testing/testing purposes
```

The use case for us/me is the protos we utilize are often highly nested. The quantity of these protos makes it impractical to define some function like

```rust
pub fn generate_my_proto(rng: &mut rng) -> MyProto { ... }
```

Would something like so be acceptable? I believe I could create the necessary functionality behind a feature flag in code generation extending what is already create for `Default::default()` but would such a feature be accepted? There's no point in writing the code for it if it forces me to fork prost in perpetuity.

**Comment by caspermeijn** [comment]
I feel like you are looking for the crate Arbitrary[https://docs.rs/arbitrary/latest/arbitrary/]

I think you could derive it automatically using prost-build. Let me know how that went, as I am interested in that as well.

It would be great if prost-types could derive arbitrary as well. That should be behind a feature flag. Are you willing to contribute that?

20 aug. 2024 15:19:11 Michael Jaquier ***@***.***>:

**Comment by mzabaluev** [comment]
[Proptest](https://crates.io/crates/proptest) is another library that can help, it provides more flexible generation strategies than Arbitrary, can shrink failure cases, and maintain a record of past regressions. proptest-derive provides the derive macro.
